README.md
## Information

<table>
<tr>
<td>Package</td><td>microphone</td>
</tr>
<tr>
<td>Description</td>
<td>microphone is a simple module that use `arecord` ALSA tools on Linux or SoX on OSX & Windows method to capture sound from a USB Microphone</td>
</tr>
</table>

## Dependencies

This library need

* ALSA tools installed on the machine (`sudo apt-get install alsa-utils`) **for Linux**
* SoX Tools installed on the machine **for Windows or OSX**

## Usage

#### Simple example

A simple example which capture sound and redirect it to stdout.

    var Mircophone = require('microphone');
    var mic = new Microphone();

    mic.on.('audio', function(audioData) {
        // just dump the recorded buffer-data to stdout
        processs.stdout.write(audioData.buffer);
        // audioData has another field
        //   audioData.stream
        // as well
    });

    mic.on('info', function(infoData) {
        process.stdout.write(infoData);
    });
    
    mic.on('audio-started', function(audioStream) {
        // you can do something with the audio-stream here (e.g. start streaming to clients)
        // it is the same, as mic.audioStream, btw
    });
    
    mic.on('audio-stopped', function() {
        // just tells you, that
        //   mic.stopCapture();
        // was called from somewhere
    });
    
    mic.startCapture();
    

## API

#### startCapture(options)

Start the process and pipe the stdout of ALSA `arecord` tool to audioStream

By default, the outputing sound are PCM WAVE format, if you want a MP3 format
pass `true` as `mp3Output` in the options passed in arguments.

(Example : `mic.startCapture({'mp3output' : true});`)

Other optional parameters in the options object include:
* alsa_format - ALSA format to use for capture (Example: 'S16_LE' See arecord --help for a list of recognized formats)
* alsa_device - ALSA device to capture from (Example: 'hw:1,0')
* alsa_addn_args - Additional arguements for ALSA (Example: ['--rate', '16000'])
* sox_format - SoX format to use for capture (Example: 'dat')
* sox_addn_args - Additional arguements for SoX (Example: ['arg1', 'arg2'])
* mp3_channels - Number of channels for the MP3 Encoder (Example: 2)
* mp3_bitDepth - Bit Depth for the MP3 Encoder (Example: 16)
* mp3_sampleRate - Sample Rate for the MP3 Encoder (Example: 44100)

#### stopCapture();

Stop the process

#### audioStream

Give the audio stream where sound captured will be pushed

#### infoStream

Give an information stream which display informations printed on `stderr` by the process

## CONTRIBUTORS

* @JiaJian *(for Windows support)*
* @guileen *(for OSX support)*
* @jrf0110
* @hanseartic *(events)*


## LICENSE

(MIT License)

Copyright (c) 2012 Vincent Saluzzo <vincentsaluzzo@me.com>

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

## Change log
[CHANGELOG.md](CHANGELOG.md)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

[//]: # (### Added)
[//]: # (### Changed)
[//]: # (### Deprecated)
[//]: # (### Removed)
[//]: # (### Fixed)
[//]: # (### Security)

## [Unreleased][unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed
### Security

## [v2.0.0] - 2015-08-17
### Added
- change log file
- events
- alsa- and SoX devices and encoding options to startCapture(options)

### Changed
- updated example in [README.md](README.md) to show events and the new way to initialize

### Deprecated
- You should not access `mic.audioStream` or `mic.infoStraem` directly anymore, but rather access the data in event-callbacks.

### Removed
### Fixed
### Security

## [v1.0.8] - 2015-01-16

## [v1.0.7] 2015-01-16
### Added
First tagged release.

[unreleased]: https://github.com/hanseartic/node-microphone/compare/v2.0.0...develop
[v2.0.0]: https://github.com/hanseartic/node-microphone/compare/v1.0.8...v2.0.0
[v1.0.8]: https://github.com/hanseartic/node-microphone/compare/v1.0.7...v1.0.8
[v1.0.7]: https://github.com/hanseartic/node-microphone/compare/061a872...v1.0.7

index.js
var EventEmitter = require('events').EventEmitter;
var util = require('util');

var isMacOrWin = require('os').type() == 'Darwin' || require('os').type().indexOf('Windows') > -1;
var spawn = require('child_process').spawn
var PassThrough = require('stream').PassThrough;
var lame = require('lame');

function Microphone () {
    EventEmitter.call(this);
    this.audioStream = new PassThrough;
    this.infoStream = new PassThrough;

    this.infoStream.on('data', function(data) {
        this.emit('info',
                  data);
    }.bind(this));
    this.audioStream.on('data', function(data) {
        this.emit('audio',
                  {
                      'stream': this.audioStream,
                      'buffer': data
                  });
    }.bind(this));
}

util.inherits(Microphone, EventEmitter);

var ps = null;

Microphone.prototype.startCapture = function(options) {
    options = options || {};
    var alsa_format = options.alsa_format || 'dat',
        alsa_device = options.alsa_device || 'plughw:1,0',
        alsa_addn_args = options.alsa_addn_args || [],
        sox_format = options.sox_format || 'dat',
        sox_addn_args = options.sox_addn_args || [],
        mp3_channels = options.mp3_channels || 2,
        mp3_bitDepth = options.mp3_bitDepth || 16,
        mp3_sampleRate = options.mp3_sampleRate || 44100;

    if(ps == null) {
        ps = isMacOrWin
        ? spawn('sox', ['-d', '-t', sox_format, '-p'].concat(sox_addn_args))
        : spawn('arecord', ['-D', alsa_device, '-f', alsa_format].concat(alsa_addn_args));

        if(options.mp3output === true) {
            var encoder = new lame.Encoder({
                channels: mp3_channels,
                bitDepth: mp3_bitDepth,
                sampleRate: mp3_sampleRate
            });

            ps.stdout.pipe(encoder);

            encoder.pipe(this.audioStream);
            ps.stderr.pipe(this.infoStream);
        } else {

            ps.stdout.pipe(this.audioStream);
            ps.stderr.pipe(this.infoStream);
        }
    }
    this.emit('audio-started', this.audioStream);
};

Microphone.prototype.stopCapture = function() {
    if(ps) {
        ps.kill();
        ps = null;
    }
    this.emit('audio-stopped');
};

module.exports = Microphone;
